### ISSUE TITLE
Genetic Programming - How to implement Op that has a state

### ISSUE THREAD

**Issue by Menoita99** [jenetics/jenetics#807]
Hello everyone!

First of all, thank you for providing such an amazing framework to work with!

I m working with jenetics GP module and I need to implement a new kind of Op that has a state.
For instance, I want to create the Op Percentage that receives a Double and do the following:

If it has not received a number before or the stored number is 0 set the value to be returned as 0.
Else set the return value to the variation between the stored number and the received number.
Then saves the received number to be used as stored number in the next evaluation.

![image](https://user-images.githubusercontent.com/36867483/139355223-f1574df3-cf2f-4866-a982-f1a85e76e551.png)

What is the best approach to implement such operation?

**Comment by jenetics** [comment]
I'm not sure that I understand your problem 100% and I'm also not sure state-full operations are the right way. 

That said, if you want to have some operation state, you can try the following:

```java
public class StateOp implements Op<Double> {
    private double state = 0;

    public StateOp(final double state) {
        this.state = state;
    }

    public StateOp() {
        this(0.0);
    }

    @Override
    public String name() {
        return String.format("state_op[%s]", state);
    }

    @Override
    public int arity() {
        return 1;
    }

    // Return a new "reset" operation with initial state.
    @Override
    public Op<Double> get() {
        return new StateOp();
    }

    @Override
    public synchronized Double apply(final Double[] args) {
        final var result = state == 0.0
            ? 0.0
            : (args[0]/state - 1.0)*100.0;
        state = args[0];
        return result;
    }
}
```
The crucial point here is the `get` method, which creates a new operation, with a reset state. This method is called whenever a new `ProgramGene` is created. For _stateless_ operations, the `get` method returns the `this` operation. As I said, I still don't think this is the right way, but it's the  most correct way I can imagine.

Let me know if this helped to create useful results ;-)

**Comment by Menoita99** [comment]
Thank you for the snippet  I really appreciate it, I will try to think of a solution without states as you have suggested.
Once I got some results I will gladly share them and explain a little better what I m trying to accomplish.

Just want to say I am a big fan of your work. I really love what you have done with jenetics and FacileJDBC.
I learned a lot by reading the source code, I didn't know java could be this awesome! I will share my experience with my colleagues and bring more people to the community!
